src/utils/geoprocessing.py

```diff
@@ -0,0 +1,16 @@
+import geopandas as gpd
+
+
+def poly_to_line(geodataframe):
+    """takes a geopandas geodataframe with polygon geometry and replaces the geometry with lines"""
+
+    # read in data from input file
+    boundary_data = geodataframe
+
+    # create new geometry column, replace original geometry column
+    boundary_data.rename_geometry('geom_old',True)
+    boundary_data['geometry'] = boundary_data.geometry.boundary
+    boundary_data = boundary_data.set_geometry('geometry')
+    boundary_data = boundary_data.drop('geom_old', 1)
```

Comment:
Ok, so a few of the mapping projects I worked on just did a line to polygon for the boundary lines.  You are right, though, this does just create line 'loops' with overlapping lines.  I can't find any SOPs for boundary lines - but I have done research into previous projects and there are a few different methods.  The most common uses ET Geowizards in ArcGIS, specifically this tool: https://www.ian-ko.com/ET_GeoWizards/UserGuide/polygonToPolylineAdvanced.htm.

This is somewhat more complex as it builds topology by properly creating nodes at intersections, de-duplicates lines, and also adds a tolerance (snapping) throughout this.  There's nothing out of the box that I can find in Geopandas/Shapely.  Have a few ideas to test.

Reply:
Hm, so I'm realizeing that this issue is actually a bit complex. I also havne't found anything out-of-the-box from GeoPandas / Shapely, but I think we could put something together.

When Matt took me through a data scramble with ArcGIS, it looked like the polygon-to-line tool created segments where each segment had one or two "parents" (i.e. countries that it borders), which is indeed very different than just spitting out each polygon boundary. And then he just took any line that had two parents as a border (having only one parent indicated a coastline).  Below is a code snippeet that I used in a different project that will output a dataframe closer to the one of ArcGIS:

```python

# Start with the standard polygon dataframe
df_borders = df_adm1.copy()

# Create a column called "borders" that contains a dictionary of intersections, where the 
# keys are the names of intersecting regions, and the values are the line gemoetry of the 
# intersection
df_borders['borders'] = df_borders.apply(lambda x: 
        {y['name_en']: x['geometry'].intersection(y['geometry'])
        for _, y in df_borders.iloc[x.name+1:].iterrows()
        if x['geometry'].intersects(y['geometry'])
    }, axis=1)

# Create a column called "new_rows" by converting each row of the "borders" column to a geo dataframe
df_borders['new_rows'] = df_borders.apply(lambda x:
     gpd.GeoDataFrame({
        'name_en_1': x['name_en'],
        'name_en_2': list(x['borders'].keys()),
        'geometry': list(x['borders'].values())
    }), axis=1
)

# Combine the rows from the "new_rows" column into a new dataframe
df_new = gpd.GeoDataFrame(columns=['name_en_1', 'name_en_2', 'geometry'])
for _, row in df_borders.iterrows():
    df_new = df_new.append(row['new_rows'], ignore_index=True)

# Convert the MultiLineStrings to just LineStrings
df_new['geometry'] = df_new['geometry'].apply(
            lambda x: ops.linemerge(x) 
            if x.geom_type == 'MultiLineString'  
            else x)
```

This will result in the dataframe that has internal boundaries only, so it would work for admin 1+.

As for admin 0, you could do something like the above, but applying the intersection check on the surrounding area boundary file. 

Finally, to get the coastline, you could just check which portion of the above admin 0 border line does not intersect the full country boundary.

Reply:
Thanks Monica - that's pretty impressive.

I've run that and it almost gives us what we want.  Took me a little while to understand it all (unpacked the lambda function to do so).  While almost what we want, I think we'll need to add in a snapping tolerance otherwise any minuscule slivers would result in no boundaries being returned.  Once I finish getting my head around that I'll add that in.

Also - I suspect the ET GeoWizards just outputs LineString as opposed to Multilinestrings as to be a proper topology network, it should split at any intersection as a node.

Is an Admin 0 Boudary without coastline used?  That's not something I've come across before.

Reply:
Ah I forgot one more thing, sometimes points can be created where three or mroe polygons meet: 
``` python
df_borders['borders'] = df_borders['borders'].apply(
              lambda x: {key: value for key, value in x.items()
               if value.geom_type != 'Point'})
```
> I've run that and it almost gives us what we want. Took me a little while to understand it all (unpacked the lambda function to do so). While almost what we want, I think we'll need to add in a snapping tolerance otherwise any minuscule slivers would result in no boundaries being returned. Once I finish getting my head around that I'll add that in.
Is this something that you found when running it, or that you foresee as an issue? I agree that it could be an issue, but I think that it's something we should be checking in the polygon DQ and repairing there. I think that e.g. admin regions within a COD should not have this issue.

> Also - I suspect the ET GeoWizards just outputs LineString as opposed to Multilinestrings as to be a proper topology network, it should split at any intersection as a node.
There is a Python package called [topojson](https://pypi.org/project/topojson/), maybe that could offfer a similar type of solution?

> Is an Admin 0 Boudary without coastline used? That's not something I've come across before.
Matt mentioned it in the [Slack thread](https://mapaction.slack.com/archives/G01148B8JES/p1592267294019100?thread_ts=1592232828.005600&cid=G01148B8JES) that Nick started. As far as I understand it, it's usefult to have separate border lines for styling, and the coastlines are needed as input to the SRTM derived coastline artefact.